### ISSUE TITLE
Support Custom Scala Dependencies

### ISSUE THREAD

**Issue by FFdhorkin** [OctoPerf/kraken#76]
**Is your feature request related to a problem? Please describe.**
We're currently using some custom Scala dependencies with gatling. I've been unable to figure out a way to get these simulations to run on Kraken. Best I can tell, Kraken actually builds the scala inside a docker image - I think that we'd need build config to be available within the Kraken interface, and then copy that into the container before building?

**Describe the solution you'd like**
Locally, we're currently using the Gatling SBT plugin, though we could switch to Maven or Gradle if we needed.

**Describe alternatives you've considered**
Gatling Frontline would probably work...

**Additional context**
Currently, our build.sbt looks like this:
```Scala
import sbt._
enablePlugins(GatlingPlugin)

// DEPENDENCIES
lazy val gatling = Seq(
    "io.gatling.highcharts" % "gatling-charts-highcharts",
    "io.gatling" % "gatling-test-framework"
).map(_ % "3.3.1" % Test)

lazy val uuid = Seq("io.jvm.uuid" %% "scala-uuid" % "0.3.1")

// lazy val requests = Seq("com.lihaoyi" %% "requests" % "0.6.0")
lazy val http = Seq("com.softwaremill.sttp" %% "core" % "1.0.5")

lazy val json4sNative = Seq("org.json4s" %% "json4s-native" % "3.7.0-M2")


// Begin build config
lazy val root = (project in file("."))
.settings(
    inThisBuild(List(
        organization := "io.forte",
        scalaVersion := "2.12.11",
        version := "0.1.0-SNAPSHOT"
    )),
    name := "forte-gatling",
    libraryDependencies ++= 
        gatling 
        ++ http
        ++ uuid
        ++ json4sNative
)

scalacOptions += "-feature"
scalacOptions += "-deprecation"
scalacOptions += "-language:postfixOps"
```

**Comment by geraldpereira** [comment]
@FFdhorkin I'm not familiar with SBT builds but from what I understand of the build.sbt file above it automatically adds `io.jvm.uuid:scala-uuid`, `com.softwaremill.sttp:core` and `org.json4s:json4s-native` to Gatling before executing it. Is this right?

In Kraken you can open the [administration UI](https://kraken.octoperf.com/administration/) and copy the dependencies jars into the `gatling/lib` folder.

**Comment by FFdhorkin** [comment]
good to know there's a manual process, but I'd prefer Kraken either handle dependency management on its own. I guess that'll work for now; I'll just need to figure out where SBT downloads its packages. As we start to have more dependencies and/or versions, though, that starts getting unwieldy. I'm OK with that for now, as it sounds like it will unblock me from testing out Kraken. Thanks!

**Comment by geraldpereira** [comment]
Merged into https://github.com/OctoPerf/kraken/issues/77

**Comment by FFdhorkin** [comment]
> @FFdhorkin I'm not familiar with SBT builds but from what I understand of the build.sbt file above it automatically adds `io.jvm.uuid:scala-uuid`, `com.softwaremill.sttp:core` and `org.json4s:json4s-native` to Gatling before executing it. Is this right?
> 
> In Kraken you can open the [administration UI](https://kraken.octoperf.com/administration/) and copy the dependencies jars into the `gatling/lib` folder.

This does not appear to have worked. Firstly, some of my .jar files threw an error message when trying to upload via the UI (though most seemed to upload):
>Error 413: Http failure response for http://_____________.elb.us-west-2.amazonaws.com/api/storage/files/set/file?path=gatling/lib: 413 Request Entity Too Large
This error disappears very, very quickly, which made copying to clipboard tricky.

Secondly, I get errors such as below when I try to run... these relate to .jar files that I successfully uploaded:
>2020-04-22 02:02:14.541  INFO 7 --- [       Thread-6] com.kraken.gatling.runner.GatlingRunner  : GATLING_HOME is set to /kraken/gatling
2020-04-22 02:02:16.049  INFO 7 --- [       Thread-6] com.kraken.gatling.runner.GatlingRunner  : 02:02:16.047 [INFO ] i.g.c.ZincCompiler$ - Compiling 9 Scala sources to /kraken/gatling/target/test-classes ...
2020-04-22 02:02:17.461  INFO 7 --- [       Thread-6] com.kraken.gatling.runner.GatlingRunner  : 02:02:17.461 [ERROR] i.g.c.ZincCompiler$ - /kraken/gatling/user-files/simulations/Postman.scala:2:33: object native is not a member of package org.json4s
2020-04-22 02:02:17.462  INFO 7 --- [       Thread-6] com.kraken.gatling.runner.GatlingRunner  : import org.json4s._, org.json4s.native.JsonMethods._
2020-04-22 02:02:17.462  INFO 7 --- [       Thread-6] com.kraken.gatling.runner.GatlingRunner  :                                 ^
2020-04-22 02:02:17.475  INFO 7 --- [       Thread-6] com.kraken.gatling.runner.GatlingRunner  : 02:02:17.475 [ERROR] i.g.c.ZincCompiler$ - /kraken/gatling/user-files/simulations/Postman.scala:10:25: object sttp is not a member of package com.softwaremill
2020-04-22 02:02:17.475  INFO 7 --- [       Thread-6] com.kraken.gatling.runner.GatlingRunner  : import com.softwaremill.sttp._
2020-04-22 02:02:17.475  INFO 7 --- [       Thread-6] com.kraken.gatling.runner.GatlingRunner  :                         ^
2020-04-22 02:02:17.839  INFO 7 --- [       Thread-6] com.kraken.gatling.runner.GatlingRunner  : 02:02:17.838 [ERROR] i.g.c.ZincCompiler$ - /kraken/gatling/user-files/simulations/Postman.scala:22:28: not found: value HttpURLConnectionBackend
2020-04-22 02:02:17.839  INFO 7 --- [       Thread-6] com.kraken.gatling.runner.GatlingRunner  :     implicit val backend = HttpURLConnectionBackend()
2020-04-22 02:02:17.839  INFO 7 --- [       Thread-6] com.kraken.gatling.runner.GatlingRunner  :                            ^
2020-04-22 02:02:17.952  INFO 7 --- [       Thread-6] com.kraken.gatling.runner.GatlingRunner  : 02:02:17.951 [ERROR] i.g.c.ZincCompiler$ - /kraken/gatling/user-files/simulations/Postman.scala:29:24: not found: value sttp

**Comment by geraldpereira** [comment]
What is the size of your jar files? The files may have been created but be empty. Can you check the uploaded file size from the Administration UI?

There is a configuration to be set on the storage backend to increase the upload size limit:
https://doc.octoperf.com/enterprise-edition/configuration/#file-upload-limit (here for Octoperf but it should be the same for Kraken). It can also be given as an environment variables `SPRING_SERVLET_MULTIPART_MAXREQUESTSIZE` and `SPRING_SERVLET_MULTIPART_MAXFILESIZE`. It will update the helm chart to support additional env variables.

Regarding the error messages, you can click on the Notifications tab (bottom right corner of the UI). From there a log icon is present on any error notification. Clicking on it will display the full stack trace.

The docker image used to run the test embed Gatling v3.3.0. It's packaged with scala 2.12. It is also a jar un the lib/ folder of Gatling but you cannot currently remove existing jars. You might be able to replace it by uploading a jar with the same name but it's a bit tricky.
